### PR TITLE
Report metrics on clone caching

### PR DESCRIPTION
**Pull request by jessitron**
This will report about cached clone hits and misses every 10 minutes, both in total and per repo.

I suspect you might want to change the implementation of the "all day" part, the logCachingDirectoryManagerMetricsAllDay method.

### CODE REVIEW

File: src/spi/clone/CachingDirectoryManager.ts [hunk 1]
```diff
@@ -60,13 +67,116 @@ export const CachingDirectoryManager: DirectoryManager = {
...
+    promisify((a, b) => setTimeout(b, a));
+
+const MetricLogFrequency = 1000 * 60 * 10; // ten minutes
+
+/* This returns a promise that never finishes, unless it throws somehow
+ */
+export function logCachingDirectoryManagerMetricsAllDay(): Promise<void> {
+    return sleepPlease(MetricLogFrequency)
+        .then(() => logger.info(CachingOfClonesMetricsCacheImpl.report().print()))
+        .then(() => logCachingDirectoryManagerMetricsAllDay());
```
Comment (ddgenome) [inline src/spi/clone/CachingDirectoryManager.ts]: Is this tail recursion cool stack-wise?

File: src/spi/clone/CachingDirectoryManager.ts [hunk 2]
```diff
@@ -60,13 +67,116 @@ export const CachingDirectoryManager: DirectoryManager = {
...
+            this.fallbackCounts[keyFor(repoId)] = 1;
+            return;
+        }
+        this.fallbackCounts[keyFor(repoId)]++;
+    }
+
+}
+
+function keyFor(repoId: RepoId) {
+    return `${repoId.owner}/${repoId.repo}`;
```
Comment (ddgenome) [inline src/spi/clone/CachingDirectoryManager.ts]: What should we do about owner/repo name collisions across GHE instances?

File: src/spi/clone/CachingDirectoryManager.ts (hunk 1 above)
Comment (jessitron) [reply src/spi/clone/CachingDirectoryManager.ts]: the promise returns immediately, it stores the then internally. so it isn't recursive on the same promise thread.

I think so.

File: src/spi/clone/CachingDirectoryManager.ts (hunk 2 above)
Comment (jessitron) [reply src/spi/clone/CachingDirectoryManager.ts]: I agree that's a thing and we're not handling it -- we're not handling GHE in tons of the client code.

**Comment by cdupuis** [comment]
@jessitron, please see #142 for an integration of caching stats with the metrics support already in the client.

**Pull request merged**